### ISSUE TITLE
ArangoDb Slowdown

### ISSUE THREAD

**Issue by normanuber** [arangodb/arangodb#2373]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ X ] 3.1
- [ ] 3.2 pre-releases: ___
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ X ] Debian .deb
- [ ] Windows, version: 
- [ ] MacOS, version:

Creating a ticket to keep track of the issue. I sent the my debugs logs in already privately.

The issue here is that after some time of read/writes, the query engine will slow down. I can restart the offending db server and things will return to normal. I assume that the offending db server is a leader for a majority of shards which is why it takes so long.

**Comment by fceller** [comment]
Under investigation.

**Comment by normanuber** [comment]
Hmm, interesting situation is when I kill two out of 3 db servers everything is still well. I start the 2 i just killed and kill the last one. Query progression starts to stall and eventually time out. Sometimes causing the coordinator to reach segfault. I can consistently reproduce this by doing what i said before.

For example in the running queries:
103275	FOR x IN @@collection LET att = APPEND(SLICE(ATTRIBUTES(x), 0, 25), "_key", true) LIMIT @offset, @count RETURN KEEP(x, att)	34.93 s	2017-03-29T16:02:40Z

and it stays this way until the query times out somewhere around 112 - 120s? The cluster can recover after starting that one downed node. Maybe this is a whole 'nother issue, but could be related to this issue where slowdowns in the dbservers could temporarily cause incremental slowdowns. (just my thought)

**Comment by neunhoef** [comment]
One general comment: AQL queries have in-memory state for performance reasons, so they cannot tolerate failover situations at all. If any cluster reconfiguration or failover happens during an AQL query, it will fail with an error, in many cases only after a timeout occurs. There is nothing we can do to change this short term, and quite possibly we do not want to change it long term either for performance reasons.

**Comment by neunhoef** [comment]
There should, however, be no crashes in the coordinator, this is a bug and it would be good if we could replicate this behaviour. You say "sometimes"? And it happens when you kill and restart DBServers during the execution of a query as shown above?

**Comment by normanuber** [comment]
I was not clear, this query is one that happened after the last death of a dbserver. Subsequent queries would fail until the dbserver was started again. The expected situation would be for subsequent queries to have gone to the 2 recently started dbservers, but I am better off creating a new issue to track this as it is separate from this ticket which refers to slowdowns.

The slowdowns/lockups are where 2 dbservers are using (~0% cpu) and a dbserver would use (~100-120% of cpu consistently in htop) and eventually would need to restart the offending dbserver in order to start processing queries.

**Comment by normanuber** [comment]
Hmm, seems that after a day of letting the stresser program (on my 1 agency, 2 db, 2 coord test cluster), I sent this to you guys, and shutting off the stresser then letting the db cool down (no write/reads), the db is slower than before.

Simple query like this in the UI query editor (100 rows): 

FOR doc IN interfaces FILTER doc.`device` == "{REDACTED}" RETURN doc
 
has time ranging in ~0.211s, ~0.425s. I restarted the first DBServer and query time dropped to ~0.078 s. Restarting the second DBserver and again query time dropped down to what was expected ~0.012 s.

I think having the stresser (or two) run for a day may reproduce the slowdown on your side.

**Comment by fceller** [comment]
We will setup a long running test here. @mpv1989

**Comment by maxkernbach** [comment]
Hi @normanuber,

ArangoDB 3.1 is EOL by now, is the issue still happening with 3.3.14?

**Comment by maxkernbach** [comment]
Closing this issue since 3.1 is EOL and no similar issue was reported with the current version.

Please feel free to comment if the problem persists using 3.3.14 or a later version and we will reopen this ticket.
